go/tools/bzltestutil/chdir/init.go

```diff
@@ -0,0 +1,138 @@
+// transitively imported packages. We relied on the linker initializing
+// packages in the order their imports appeared in source, so we import
+// bzltestutil (and transitively, this package) from the generated test main
+// before other packages.
+//
+// In Go 1.21, the package initialization order was clarified, and the
+// linker implementation was changed. See
+// https://go.dev/ref/spec#Program_initialization or
+// https://go.dev/doc/go1.21#language.
+//
+// > Given the list of all packages, sorted by import path, in each step the
+// > first uninitialized package in the list for which all imported packages
+// > (if any) are already initialized is initialized. This step is repeated
+// > until all packages are initialized.
+//
+// To ensure this package is initialized before user code without injecting
+// edges into the dependency graph, we implement the following hack:
+//
+// 1. Add the prefix '+initfirst/' to this package's path with the 'importmap'
+//    attribute. '+' is the first allowed character that sorts higher than
+//    letters. Because we're using 'importmap' and not 'importpath', this
+//    package may be imported in .go files without the prefix.
+// 2. Put this init function in a separate package that only imports "os".
+//    Previously, this function was in bzltestutil, but bzltest util imports
+//    several other std packages may be get initialized later. For example,
+//    the "sync" package is initialized after a user package named
+//    "github.com/a/b" that only imports "os", and because bzltestutil imports
+//    "sync", it would get initialized even later. A user package that imports
+//    nothing may still be initialized before "os", but we assume "os"
+//    is needed to observe the current directory.
+package chdir
+
+// This package should import nothing other than "os"
+// and packages imported by "os".
+import (
+	"os"
+	"runtime"
+)
+
+var (
+	// Initialized by linker.
+	RunDir string
+
+	// Initial working directory.
+	TestExecDir string
+)
+
+func init() {
+	var err error
+	TestExecDir, err = os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	// Check if we're being run by Bazel and change directories if so.
+	// TEST_SRCDIR and TEST_WORKSPACE are set by the Bazel test runner, so that makes a decent proxy.
+	testSrcDir, hasSrcDir := os.LookupEnv("TEST_SRCDIR")
+	testWorkspace, hasWorkspace := os.LookupEnv("TEST_WORKSPACE")
+	if hasSrcDir && hasWorkspace && RunDir != "" {
+		abs := RunDir
+		if !filepathIsAbs(RunDir) {
+			abs = filepathJoin(testSrcDir, testWorkspace, RunDir)
+		}
+		err := os.Chdir(abs)
+		// Ignore the Chdir err when on Windows, since it might have have runfiles symlinks.
+		// https://github.com/bazelbuild/rules_go/pull/1721#issuecomment-422145904
+		if err != nil && runtime.GOOS != "windows" {
+			panic("could not change to test directory: " + err.Error())
+		}
+		if err == nil {
+			os.Setenv("PWD", abs)
+		}
+	}
+}
+
+func filepathIsAbs(path string) bool {
```

Comment:
Is this copied from the Go stdlib sources? If so, please add a link.

Reply:
No, this is a more minimal hack, same with `filepathJoin`. I've expanded the comment to explain its limitations.